eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs

```diff
@@ -180,54 +281,63 @@ instance DecCBOR Tag where
+encodeScript :: AlonzoEraScript era => AlonzoScript era -> Encode 'Open (AlonzoScript era)
 encodeScript = \case
   TimelockScript i -> Sum TimelockScript 0 !> To i
-  PlutusScript (Plutus PlutusV1 s) -> Sum (PlutusScript . Plutus PlutusV1) 1 !> To s
-  PlutusScript (Plutus PlutusV2 s) -> Sum (PlutusScript . Plutus PlutusV2) 2 !> To s
-  PlutusScript (Plutus PlutusV3 s) -> Sum (PlutusScript . Plutus PlutusV3) 3 !> To s
+  PlutusScript plutusScript -> withPlutusScript plutusScript $ \plutus@(Plutus pb) ->
+    case plutusSLanguage plutus of
+      SPlutusV1 -> Sum (PlutusScript . fromJust . mkPlutusScript . Plutus @'PlutusV1) 1 !> To pb
+      SPlutusV2 -> Sum (PlutusScript . fromJust . mkPlutusScript . Plutus @'PlutusV2) 2 !> To pb
+      SPlutusV3 -> Sum (PlutusScript . fromJust . mkPlutusScript . Plutus @'PlutusV3) 3 !> To pb
 
-instance Era era => DecCBOR (Annotator (AlonzoScript era)) where
-  decCBOR = decode (Summands "Alonzo Script" decodeScript)
+instance AlonzoEraScript era => DecCBOR (Annotator (AlonzoScript era)) where
+  decCBOR = decode (Summands "AlonzoScript" decodeScript)
     where
-      decodeAnnPlutus lang =
-        Ann (SumD $ PlutusScript . Plutus lang) <*! Ann (D (guardPlutus lang >> decCBOR))
+      decodeAnnPlutus slang =
+        Ann (SumD PlutusScript) <*! Ann (D (decodePlutusScript slang))
       {-# INLINE decodeAnnPlutus #-}
       decodeScript :: Word -> Decode 'Open (Annotator (AlonzoScript era))
       decodeScript = \case
         0 -> Ann (SumD TimelockScript) <*! From
-        1 -> decodeAnnPlutus PlutusV1
-        2 -> decodeAnnPlutus PlutusV2
-        3 -> decodeAnnPlutus PlutusV3
+        1 -> decodeAnnPlutus SPlutusV1
+        2 -> decodeAnnPlutus SPlutusV2
+        3 -> decodeAnnPlutus SPlutusV3
         n -> Invalid n
       {-# INLINE decodeScript #-}
   {-# INLINE decCBOR #-}
 
 -- | Test that every Alonzo script represents a real Script.
 --     Run deepseq to see that there are no infinite computations and that
 --     every Plutus Script unflattens into a real PV1.Script
-validScript :: ProtVer -> AlonzoScript era -> Bool
+validScript :: AlonzoEraScript era => ProtVer -> Script era -> Bool
 validScript pv script =
-  case script of
-    TimelockScript sc -> deepseq sc True
-    PlutusScript (Plutus lang (BinaryPlutus bytes)) ->
-      let deserialiseScript =
-            case lang of
-              PlutusV1 -> PV1.deserialiseScript
-              PlutusV2 -> PV2.deserialiseScript
-              PlutusV3 -> PV3.deserialiseScript
-          eWellFormed :: Either PV1.ScriptDecodeError PV1.ScriptForEvaluation
-          eWellFormed = deserialiseScript (transProtocolVersion pv) bytes
-       in isRight eWellFormed
+  case toPlutusScript script of
+    Just plutusScript -> isValidPlutusScript (pvMajor pv) plutusScript
+    Nothing ->
+      case getNativeScript script of
+        Just timelockScript -> deepseq timelockScript True
+        Nothing -> False -- Impossible case
```

Comment:
Maybe we should throw an error here instead? That's how we've usually done it for impossible cases.

Reply:
Sure, I can switch it to `error`.

Reply:
Done.